runtime/skiplist.c

```diff
@@ -0,0 +1,214 @@
+   1 with probability 3/16, 2 with probability 3/64, etc.
+   We use a simple linear congruential PRNG (see Knuth vol 2) instead
+   of random(), because we need exactly 32 bits of pseudo-random data
+   (i.e. 2 * (NUM_LEVELS - 1)).  Moreover, the congruential PRNG
+   is faster and guaranteed to be deterministic (to reproduce bugs). */
+
+static uint32_t random_seed = 0;
+
+static int random_level(void)
+{
+  uint32_t r;
+  int level = 0;
+
+  /* Linear congruence with modulus = 2^32, multiplier = 69069
+     (Knuth vol 2 p. 106, line 15 of table 1), additive = 25173. */
+  r = random_seed = random_seed * 69069 + 25173;
+  /* Knuth (vol 2 p. 13) shows that the least significant bits are
+     "less random" than the most significant bits with a modulus of 2^m,
+     so consume most significant bits first */
+  while ((r & 0xC0000000U) == 0xC0000000U) { level++; r = r << 2; }
+  CAMLassert(level < NUM_LEVELS);
+  return level;
+}
+
+/* Initialize a skip list */
+
+void caml_skiplist_init(struct skiplist * sk)
+{
+  sk->forward[0] = NULL;
+  sk->level = 0;
+}
+
+/* Search a skip list */
+
+int caml_skiplist_find(struct skiplist * sk, uintnat key, uintnat * data)
+{
+  int i;
+  struct skipcell * e, * f;
+
+  e = (struct skipcell *) sk;
+  for (i = sk->level; i >= 0; i--) {
+    while (1) {
+      f = e->forward[i];
+      if (f == NULL || f->key >= key) break;
+      e = f;
+    }
+  }
+  e = e->forward[0];
+  if (e != NULL && e->key == key) {
+    *data = e->data;
+    return 1;
+  } else {
+    return 0;
+  }
+}
+
+int caml_skiplist_find_below(struct skiplist * sk, uintnat k,
+                             uintnat * key, uintnat * data)
+{
+  int i;
+  struct skipcell * e, * f;
+
+  e = (struct skipcell *) sk;
+  for (i = sk->level; i >= 0; i--) {
+    while (1) {
+      f = e->forward[i];
+      if (f == NULL || f->key >= k + 1) break;
+      e = f;
+    }
+  }
+  if (e == (struct skipcell *) sk) {
+    return 0;
+  } else {
+    *key = e-> key; *data = e->data; return 1;
+  }
+}
+
+/* Insertion in a skip list */
+
+int caml_skiplist_insert(struct skiplist * sk,
+                         uintnat key, uintnat data)
+{
+  struct skipcell * update[NUM_LEVELS];
+  struct skipcell * e, * f;
+  int i, new_level;
+
+  CAMLassert(0 <= sk->level && sk->level < NUM_LEVELS);
+
+  /* Init "cursor" to list head */
+  e = (struct skipcell *) sk;
+  /* Find place to insert new node */
+  for (i = sk->level; i >= 0; i--) {
+    while (1) {
+      f = e->forward[i];
+      if (f == NULL || f->key >= key) break;
+      e = f;
+    }
+    update[i] = e;
+  }
+  e = e->forward[0];
+  /* If already present, update data */
+  if (e != NULL && e->key == key) {
+    e->data = data;
+    return 1;
+  }
+  /* Insert additional element, updating list level if necessary */
+  new_level = random_level();
+  if (new_level > sk->level) {
+    for (i = sk->level + 1; i <= new_level; i++)
+      update[i] = (struct skipcell *) sk;
```

Comment:
It seems like nothing guarantees here that `sk->forward[i] == NULL`.
If this is supposed to be an invariant of the data structure, then it does not hold after `caml_skiplist_init`.
If this guarantee is not supposed to hold at this point, then how do we make sure that the skip list at level `i` finishes with `NULL`?

Reply:
Very good point, thanks.  It is an invariant that forward pointers above the current level are NULL.  But `caml_skiplist_init` did not establish this invariant.  This is fixed in 881a2c0a0.